src/state.rs

```diff
@@ -41,7 +40,29 @@ use hole_punching::HolePunchServer;
 use util::ip_from_socketaddr;
 use util;
 
-pub type Closure = Box<FnBox(&mut State) + Send>;
+// Closure is a wapper around boxed closures that tries to work around the fact
+// that it is not possible to call Box<FnOnce> in the current stable rust.
+// The idea is to wrap the FnOnce in something that implements FnMut using some
+// dirty tricks, because Box<FnMut> is fine to call.
+//
+// This workaround can be removed once FnBox becomes stable or Box<FnOnce>
+// becomes usable.
+pub struct Closure(Box<FnMut(&mut State) + Send>);
+
+impl Closure {
+    pub fn new<F : FnOnce(&mut State) + Send + 'static>(f: F) -> Closure {
+        let mut f = Some(f);
+        Closure(Box::new(move |state: &mut State| {
```

Comment:
Yeah. Basically we need a closure that implements `FnOnce` to be stored in a `Box`. This is not usable in the current stable rust, because you can't call it back. That's why `FnBox` was introduced, but it is behind a unstable feature gate, so we can't use it either. The workaround is basically to use `Box<FnMut>` instead, but to be able to store `FnOnce`'s in it, I need to put that `FnOnce` into `Option` and then take it out when calling it. There is a lengthy and detailed discussion about this issue here: https://github.com/rust-lang/rust/issues/28796

The `Closure` struct is there just to wrap all this hackery in a convenient interface.